include/albatross/src/indexing/traits.hpp

```diff
@@ -0,0 +1,195 @@
+  static TypePair<std::true_type, ReturnType> test(C *);
+  template <typename> static TypePair<std::false_type, void> test(...);
+
+public:
+  static constexpr bool is_defined = decltype(test<T>(0))::first_type::value;
+  using return_type = typename decltype(test<T>(0))::second_type;
+};
+
+/*
+ * Here we only care about the `is_defined` trait.
+ */
+template <typename T, typename... Args> struct can_be_called_with {
+  static constexpr bool value = callable_traits<T, Args...>::is_defined;
+};
+
+template <typename T, typename... Args>
+class can_be_called_with_const_ref
+    : public can_be_called_with<const T, typename const_ref<Args>::type...> {};
+
+/*
+ * Stores the return type resulting from calling T with Args, if the
+ * call isn't valid the resulting type will be void.
+ */
+template <typename T, typename... Args> struct return_type_when_called_with {
+  using type = typename callable_traits<T, Args...>::return_type;
+};
+
+/*
+ * A GrouperFunction is a function which takes a single const ref argument
+ * and returns a non-void type which will end up being the key used in group by.
+ */
+template <typename GrouperFunction, typename ValueType>
+struct grouper_return_type
+    : public return_type_when_called_with<GrouperFunction,
+                                          typename const_ref<ValueType>::type> {
+};
+
+template <typename GrouperFunction, typename ValueType>
+struct group_key_is_valid {
+  static constexpr bool value = is_valid_map_key<
+      typename grouper_return_type<GrouperFunction, ValueType>::type>::value;
+};
+
+template <typename GrouperFunction, typename ValueType>
+struct is_valid_grouper {
+
+  static constexpr bool value =
+      can_be_called_with_const_ref<GrouperFunction, ValueType>::value &&
+      group_key_is_valid<GrouperFunction, ValueType>::value;
+};
+
+/*
+ * An ApplyFunction is a function which takes a key value pair and
+ * returns a new type, they key type needs to remain unchanged but
+ * the value can be modified.
+ */
+template <typename ApplyFunction, typename KeyType, typename ArgType>
+struct key_value_apply_return_type
+    : public return_type_when_called_with<
+          ApplyFunction, typename const_ref<KeyType>::type, ArgType> {};
+
+template <typename ApplyFunction, typename ArgType>
+struct value_only_apply_return_type
+    : public return_type_when_called_with<ApplyFunction, ArgType> {};
+
+template <typename ApplyFunction, typename KeyType, typename ArgType>
+struct is_valid_key_value_apply_function
+    : public can_be_called_with_const_ref<
+          ApplyFunction, typename const_ref<KeyType>::type, ArgType> {};
+
+template <typename ApplyFunction, typename ArgType>
+struct is_valid_value_only_apply_function
+    : public can_be_called_with_const_ref<ApplyFunction, ArgType> {};
+
+template <typename ApplyFunction, typename KeyType, typename ArgType>
+struct is_valid_index_apply_function
+    : public can_be_called_with_const_ref<
```

Comment:
Any reason why you jump between inheriting from other traits and explicitly defining the `value`/`type` members? They should be equivalent, unless there's additional members you don't want to public inherit, but I'm not seeing any cases of that.

Reply:
No specific reason ... I do plan on overhauling the traits in a follow up PR so I'll leave this for now, then aim for consistency in the follow up.